Improved Std-Vector Operations

oschulz opened the pull request:
Currently includes the following std-vector operations:

* A `getindex` specialization for `StdVector` that returns values instead of references for vectors of primitive C++ types
* `setindex!` for `StdVector`
* A high performance `copy!` between `StdVector` and `AbstractVector` for primitive C++ types. With the `setindex!` above, the default Julia `copy!` implementation works, but is very slow in comparison.

@Keno, would something along these lines be acceptable?

Keno commented on src/std.jl:
```diff
@@ -65,6 +71,32 @@ function Base.filter!(f, a::StdVector)
+
+    const src_data = @cxx src->data()
+    @inbounds @simd for i in 1:src_length
+        dest[i] = unsafe_load(src_data, i)
```
unsafe_copy!

oschulz replied:
With unsafe_load, this won't work for things like SparseVector, though, right (the current version does)? Performance wise (benchmarked with array sizes 32, 128, 1024 and 10240) there doesn't seem to be any gain with `unsafe_copy!`. The resulting machine code of the current version with the loop looks nicely vectorized (for `Vector`, not for `SparseVector`, of course).

I'm far from being an expert on `unsafe_...`, though - is there another reason why `unsafe_copy!` might be preferable? If so, should I keep the generic version for `AbstractVector` like now and add a specialized version for `DenseVector` using `unsafe_copy!`?

oschulz replied:
I realized that the version with the loop won't work for the new custom-indexed arrays. I added the necessary code for index-shifts, but it decreases performance for mid-sized array. So I guess `DenseVector` is best served by `unsafe_copy!`, and the generic `AbstractVector`case (e.g. `SparseVector`) by a loop that can also handle custom indices.

oschulz replied:
Admittedly, a high-performance copy between `StdVector` and `SparseVector` or similar may not be a very common use case.

Keno commented on src/std.jl:
```diff
@@ -65,6 +71,32 @@ function Base.filter!(f, a::StdVector)
+
+    const dest_data = @cxx dest->data()
+    @inbounds @simd for i in 1:src_length
+        unsafe_store!(dest_data, src[i], i)
```
likewise

Keno commented on src/std.jl:
```diff
@@ -35,7 +35,13 @@ end
 Base.done(it::StdVector,i) = i >= length(it)
+
 Base.getindex(it::StdVector,i) = icxx"($(it))[$i];"
+Base.getindex{T<:Cxx.CxxBuiltinTs}(it::StdVector{T}, i::Int) = icxx"auto x = ($(it))[$i]; x;"
```
I'd like to see tests for these.

oschulz replied:
Speaking of tests, where should tests for std.jl go ("tests/misc.jl", a new one, ...)?

oschulz commented:
Sorry for the monologue, Keno. :-)

I'll revise and extend this PR and push a new version soon.

oschulz commented:
Revised, extended and rebased. This now covers the following improvements and additions for StdVector:

* Specialized version of `getindex` for primitive-type std-vectors, returning the value instead of the reference.

* `setindex!` - several versions are necessary to cover types with/without Cxx.CppValue and conversion.

* `linearindices` and `checkbounds` and bounds-checking for `getindex` and `setindex!`.

* `resize!`

* `copy!(dest::StdVector, src)` (Julia already covers the reverse direction).

* High performance `copy!` between StdVector and DenseVector for vectors with primitive element types (based on `unsafe_copy!`).

* `convert` between StdVector and AbstractVector, based on `copy!`.

I also added `convert` for String and StdString (both directions). In conjunction with the other additions, string-vector `copy!`/`convert`/`setindex!` works too:

    jl_strs1 = ["foo", "blah"]
    cxx_strs = convert(cxxt"std::vector<std::string>", jl_strs1)
    cxx_strs[0] = "baz"
    jl_strs2 = convert(Vector{String}, cxx_strs)
    jl_strs2 == ["baz", "blah"]

@Keno: I do realize that I'll have to add tests for all of this - before I dive into that, is this acceptable in general? Also, into which file should tests go?

Keno commented:
Yes, I think this is fine in general. The only concern is the additional start-up time, but that's really my problem, since it's pretty bad already. As for test file, I'd just create an additional `std.jl` test file. It doesn't really go into any of the existing ones.

oschulz commented:
Thanks, Keno, I'll start writing tests.

oschulz commented:
Sorry, noticed that `copy!` for `StdVector` and `DenseVector` was missing methods with offsets. Hope I'll finally get around to writing tests now. :-)

oschulz commented:
I've re-designed the `copy!` and `convert` support for `StdVector`: New methods of `unsafe_wrap` for `StdVector` return a `DenseArray`, either a `WrappedCppObjArray` or a `WrappedCppPrimArray`. Both are bitstype (so that they'll be stack-allocated). `copy!` and `convert` are forwarded to the wrapper and the standard Julia array methods take care of all the rest. Performance is very good, and this way things are more generic (and maybe also more robust) than before.

Test implementation will follow.

oschulz commented:
Added tests. @Keno, I hope you like it - could you give this a final review? From my side, it's ready to merge. Should I squash / clean up the history a bit more or not?

Keno commented:
Up to you. I only require clean histories for mature packages, since I myself am not particularly careful about that for me less mature packages ;). This package is somewhere on the edge. I'll try to get around to reviewing this soon. If I don't in a few days, please bump this thread.

oschulz commented:
I did clean the history up a bit, after all. I like clean histories as well. ;-)

oschulz commented:
@keno, I just discovered that there's a problem with `std::vector<bool>` - I forgot about the fact that it's not a "normal" STL-vector and usually stored as a bit array. So this PR shouldn't be merged yet - I'll find a fix and add a few tests.

oschulz commented:
Ok, problems with `std::vector<bool>` are fixed. Performance for bool vectors won't be great, as the C++-standard doesn't define the memory layout (so we need to go via C++ for each element access).

@Keno bump to you for review/merge. :-)

oschulz commented:
@Keno : Bump. :-)

oschulz commented:
@Keno: Bump. :-)

Keno commented:
Sorry, looking over this now.

Keno commented on src/std.jl:
```diff
@@ -15,6 +16,8 @@ typealias StdMap{K,V} cxxt"std::map<$K,$V>"
 unsafe_string(str::Union{StdString,StdStringR}) = unsafe_string((@cxx str->data()),@cxx str->size())
 String(str::Union{StdString,StdStringR}) = unsafe_string(str)
+Base.convert(::Type{String}, x::Union{StdString,StdStringR}) = String(x)
+Base.convert(StdString, x::AbstractString) = icxx"std::string s($(pointer(x)), $(sizeof(x))); s;"
```
`icxx"std::string{$(pointer(x)),$(sizeof(x))};"` should work.

oschulz replied:
Hm, it does. I started doing this because when I tried to use

    result = icxx"std::vector<$T>($n);"`

in

    function Base.convert{T}(::Type{cxxt"std::vector<$T>"}, x::AbstractArray)

 `result` was void, so I used

    result = icxx"std::vector<$T> v($n); v;"

instead, and then adopted it in other places where I needed a non-aggregate constructor.

Why `std::string{...}` though, and not `std::string(...)`? I just checked, both seem to work. However,  the curly-braces version actually requires `$(Csize_t(sizeof(x)))`, otherwise I get an

    error: non-constant-expression cannot be narrowed from type 'long' to 'size_type' (aka 'unsigned long') in initializer list

whereas the round-braces version is happy with `$(sizeof(x))`. Is this to be expected?

Keno commented on src/std.jl:
```diff
@@ -32,14 +35,35 @@ end
+@inline Base.getindex(v::StdVector,i) = (@boundscheck checkbounds(v, i); icxx"($(v))[$i];")
+@inline Base.getindex{T<:Cxx.CxxBuiltinTs}(v::StdVector{T}, i) = (@boundscheck checkbounds(v, i); icxx"$T x = ($(v))[$i]; x;")
+
+@inline Base.setindex!{T}(v::StdVector{T}, val::T, i::Integer) =
```
`setindex!` returns the modified vector

oschulz replied:
Ah, sorry, my bad - I only looked at `v[i] = ...`, where Julia overrides the return value of `setindex!` anyhow. I didn't realize that it wasn't an exact equivalent to `setindex!`. Would it be better to use

````
@inline Base.setindex!{T}(v::StdVector{T}, val::T, i::Integer) =
    (@boundscheck checkbounds(v, i); icxx"($(v))[$i] = $val; $v;")
````

or

````
@inline Base.setindex!{T}(v::StdVector{T}, val::T, i::Integer) =
    (@boundscheck checkbounds(v, i); icxx"($(v))[$i] = $val; void();"; v)
````

performance-wise?

Keno commented on src/std.jl:
```diff
@@ -65,6 +92,89 @@ function Base.filter!(f, a::StdVector)
 
+typealias CxxBuiltinVecTs Union{Float32,Float64,Int16,Int32,Int64,Int8,UInt16,UInt32,UInt64,UInt8}
+
+Base.unsafe_wrap{T}(::Type{DenseArray}, v::StdVector{T}) = WrappedCppObjArray(pointer(v), length(v))
```
I'd prefer the type definition for this before the methods that use it.

oschulz replied:
Uh, I'm probably misunderstanding something here. `typealias CxxBuiltinVecTs` is located before the methods that use it. But I guess that's not what you mean?

Keno commented on src/std.jl:
```diff
@@ -65,6 +92,89 @@ function Base.filter!(f, a::StdVector)
+    setindex!(A, convert(T, val), i)
+
+
+immutable WrappedCppObjArray{T, CVR} <: WrappedCppDenseValues{T}
```
I'm concerned that these don't keep a reference to the array they're wrapping. If you're not careful, the underlying storage will be deallocated.

oschulz replied:
Well, my aim was to make this very fast, and avoid memory allocation and GC. I often have use cases where the same vector is copied many times (and modified on the C++ side between copies) -  so this way I can avoid mem alloc and GC almost completely. I thought it should be Ok not to keep a reference here:

* `WrappedCppObjArray` and `WrappedCppPrimArray` are not exported - a user will typically get them from a call to `unsafe_wrap`. It's my understanding that `unsafe_wrap` implies that the user is responsible from preventing the object being unsafely wrapped from getting GC'ed. The wrapper are definitely intended to be very short-lived.

* When used internally by `copy!` and `convert` the object should never be GC'ed while the function is running and it's argument is still in scope, right?

oschulz commented:
Sorry, need a bit of guidance (see my comments). ;-)

oschulz commented:
Bump. ;-)

oschulz commented:
Found one more thing I have to fix.

oschulz commented:
Fixed one more thing

Keno commented:
LGTM, Thanks!

oschulz commented:
Thanks!
